File: internal/egress/allowlist.go
// Package egress is the allowlisting proxy a harness run reaches the internet
// through (D12.6, D12.10).
//
// A harness container has no route out. Its only path to the network is this
// proxy, sitting on a per-run internal Podman network, and the proxy answers one
// question: may this run reach this host on this port. Absence of an allowlist
// entry is deny, which is invariant 1 applied to egress.
//
// The trifecta rule (D10) is why this exists. A harness has private data through
// MCP, untrusted content from files and the web, and egress. Egress is the leg
// that is cheap to bound, so it is bounded per run rather than per deployment.
package egress

import (
	"errors"
	"fmt"
	"net"
	"strconv"
	"strings"
)

// DefaultPorts are the ports an entry without one permits. Deliberately short:
// an agent that needs 8443 should say 8443.
var DefaultPorts = []int{80, 443}

// Rule is one allowlist entry.
type Rule struct {
	// Host is a hostname, an IP literal, or a "*." wildcard.
	Host string

	// Ports the rule permits. Empty means DefaultPorts.
	Ports []int

	// wildcard is set for "*.example.com" forms.
	wildcard bool

	// allowPrivate lets THIS rule reach a non-public address.
	//
	// Per rule, not per allowlist. The global flag it replaces meant that
	// allowing one LAN printer widened the SSRF guard for every other entry,
	// so a run that legitimately needed one private host also got permission
	// to follow api.example.com to 169.254.169.254. The narrowest scope that
	// does the job is the rule that asked for it.
	allowPrivate bool

	// ip is set when Host parsed as an IP literal, so an explicit address entry
	// can be matched without a DNS round trip.
	ip net.IP
}

// Allowlist decides what a run may reach. The zero value denies everything,
// which is the point: a proxy that starts up empty is safe, not broken.
type Allowlist struct {
	rules []Rule

	// AllowPrivateDestinations permits RFC1918, loopback, link-local and ULA
	// destinations. Off by default.
	//
	// This is the DNS-rebinding and SSRF control. Without it, allowlisting
	// `metrics.example.com` is enough to reach 169.254.169.254 or the host's
	// own services, because the attacker controls what the name resolves to.
	// Turning it on is for LAN targets a run legitimately needs (tools.md's
	// local browser driver), and it is a deliberate widening.
	AllowPrivateDestinations bool
}

// ParseAllowlist builds an allowlist from entries.
//
// Accepted forms:
//
//	example.com            host on ports 80 and 443
//	example.com:8443       host on one port
//	*.example.com          subdomains of example.com, NOT example.com itself
//	*.example.com:8443     subdomains on one port
//	192.0.2.10             an address literal
//	192.0.2.10:9000        an address literal on one port
//
// An empty list is valid and denies everything.
func ParseAllowlist(entries []string) (*Allowlist, error) {
	list := &Allowlist{}

	for _, raw := range entries {
		entry := strings.TrimSpace(raw)
		if entry == "" {
			continue
		}

		rule, err := parseRule(entry)
		if err != nil {
			return nil, fmt.Errorf("allowlist entry %q: %w", raw, err)
		}
		list.rules = append(list.rules, rule)
	}

	return list, nil
}

// PrivatePrefix marks an entry as permitted to resolve to a non-public address.
//
//	private:printer.home.example.com
//
// Spelled out rather than inferred, because "this name is allowed to point
// inside the network" is exactly the sentence someone should have to write
// down.
const PrivatePrefix = "private:"

func parseRule(entry string) (Rule, error) {
	var allowPrivate bool
	if rest, ok := strings.CutPrefix(entry, PrivatePrefix); ok {
		allowPrivate = true
		entry = strings.TrimSpace(rest)
		if entry == "" {
			return Rule{}, fmt.Errorf("no host after the %q prefix", PrivatePrefix)
		}
	}

	host := entry
	var ports []int

	// SplitHostPort would reject a bare hostname and mangle a bare IPv6, so
	// only split when there is a port to split off.
	if h, p, err := net.SplitHostPort(entry); err == nil {
		port, convErr := strconv.Atoi(p)
		if convErr != nil {
			return Rule{}, fmt.Errorf("port %q is not a number", p)
		}
		if port < 1 || port > 65535 {
			return Rule{}, fmt.Errorf("port %d is out of range", port)
		}
		host = h
		ports = []int{port}
	} else if strings.HasPrefix(entry, "[") {
		// A bracketed IPv6 with no port: strip the brackets.
		host = strings.TrimSuffix(strings.TrimPrefix(entry, "["), "]")
	}

	host = strings.ToLower(strings.TrimSuffix(host, "."))
	if host == "" {
		return Rule{}, errors.New("no host")
	}

	rule := Rule{Ports: ports, allowPrivate: allowPrivate}

	switch {
	case host == "*":
		// An allowlist that allows everything is almost certainly a mistake,
		// and if it is not, it should be spelled out host by host.
		return Rule{}, errors.New(`"*" would allow every host; list them explicitly`)

	case strings.HasPrefix(host, "*."):
		suffix := host[1:] // keep the leading dot
		if strings.ContainsAny(suffix[1:], "*") || suffix == "." {
			return Rule{}, errors.New("malformed wildcard")
		}
		rule.Host = suffix
		rule.wildcard = true

	case strings.Contains(host, "*"):
		return Rule{}, errors.New("a wildcard is only supported as a leading \"*.\" label")

	default:
		rule.Host = host
		rule.ip = net.ParseIP(host)
		// Naming a private address literally IS the opt-in. Writing
		// 192.168.1.50 in an allowlist and having it silently never match was
		// the other half of this bug: the entry looked effective and was inert,
		// which is the worst kind of configuration.
		if rule.ip != nil && isPrivate(rule.ip) {
			rule.allowPrivate = true
		}
	}

	return rule, nil
}

// Rules returns the parsed rules, for logging and tests.
func (a *Allowlist) Rules() []Rule { return a.rules }

// Empty reports whether the allowlist permits nothing.
func (a *Allowlist) Empty() bool { return a == nil || len(a.rules) == 0 }

// Permits reports whether host:port is allowed by name. It says nothing about
// where the name resolves; [Allowlist.PermitsAddr] is the other half.
func (a *Allowlist) Permits(host string, port int) bool {
	_, ok := a.Match(host, port)
	return ok
}

// Match returns the rule that permits host:port, if any.
//
// The rule matters to the second check: whether a resolved address may be
// dialled is a question about the entry that allowed the name, not about the
// allowlist as a whole.
func (a *Allowlist) Match(host string, port int) (Rule, bool) {
	if a == nil {
		return Rule{}, false
	}

	host = strings.ToLower(strings.TrimSuffix(host, "."))
	// A bracketed literal reaches here from a request URI.
	host = strings.TrimSuffix(strings.TrimPrefix(host, "["), "]")

	for _, rule := range a.rules {
		if !rule.permitsPort(port) {
			continue
		}
		if rule.matchesHost(host) {
			return rule, true
		}
	}
	return Rule{}, false
}

func (r Rule) permitsPort(port int) bool {
	if len(r.Ports) == 0 {
		for _, p := range DefaultPorts {
			if p == port {
				return true
			}
		}
		return false
	}
	for _, p := range r.Ports {
		if p == port {
			return true
		}
	}
	return false
}

func (r Rule) matchesHost(host string) bool {
	if r.wildcard {
		// "*.example.com" covers sub.example.com and deeper, and deliberately
		// not example.com. Allowing the apex by implication is how an allowlist
		// ends up broader than the person writing it believed.
		return strings.HasSuffix(host, r.Host)
	}
	if r.ip != nil {
		if candidate := net.ParseIP(host); candidate != nil {
			return r.ip.Equal(candidate)
		}
		return false
	}
	return host == r.Host
}

// DeniedError is a refusal by policy, as opposed to a network failure.
//
// The distinction is load-bearing: a policy denial is a 403 and means the
// allowlist did its job, while an unreachable upstream is a 502 and means
// something is broken. Collapsing them into one status made a DNS failure
// inside the proxy look exactly like a correctly-enforced allowlist, which cost
// an hour the first time it happened.
type DeniedError struct {
	Reason string
}

func (e *DeniedError) Error() string { return e.Reason }

// PermitsAddr reports whether a resolved address may be dialled under a rule.
//
// Separate from [Allowlist.Permits] because the two failures are different: a
// name that is not on the list is a misconfiguration, and a name on the list
// that resolves somewhere private is an attack or an accident that looks like
// one.
//
// It takes the matching rule rather than consulting a list-wide flag. Under the
// old shape, allowing one LAN host widened the guard for every entry, so a run
// that needed a printer also got permission to follow a public name to the
// metadata service.
func (a *Allowlist) PermitsAddr(rule Rule, ip net.IP) error {
	if a == nil {
		return &DeniedError{Reason: "no allowlist configured"}
	}
	if !isPrivate(ip) {
		return nil
	}
	// The blunt instrument stays, for a deployment that genuinely wants it,
	// and is now the exception rather than the mechanism.
	if a.AllowPrivateDestinations || rule.allowPrivate {
		return nil
	}
	return &DeniedError{Reason: fmt.Sprintf("destination %s is not a public address", ip)}
}

// isPrivate covers everything a run has no business reaching by resolving a
// public name: loopback, RFC1918, link-local (which is where cloud metadata
// services live), ULA, and the unspecified address.
func isPrivate(ip net.IP) bool {
	if ip == nil {
		return true
	}
	if ip.IsLoopback() || ip.IsPrivate() || ip.IsUnspecified() ||
		ip.IsLinkLocalUnicast() || ip.IsLinkLocalMulticast() ||
		ip.IsInterfaceLocalMulticast() || ip.IsMulticast() {
		return true
	}
	// No IPv4-mapped re-check here, deliberately. It looks necessary and is
	// dead code: net.IP.Equal treats a 4-byte address and its 4-in-6 form as
	// equal, and IsLoopback, IsPrivate and the link-local checks all normalise
	// internally, so ::ffff:127.0.0.1 is already caught above. A re-check that
	// cannot fire, with a comment claiming it is the guard, is worse than no
	// comment at all.
	// 100.64.0.0/10, carrier-grade NAT, also used by container runtimes.
	if v4 := ip.To4(); v4 != nil && v4[0] == 100 && v4[1] >= 64 && v4[1] <= 127 {
		return true
	}
	return false
}

// String renders a rule the way it would be written in configuration.
func (r Rule) String() string {
	host := r.Host
	if r.wildcard {
		host = "*" + r.Host
	}
	if len(r.Ports) == 0 {
		return host
	}
	parts := make([]string, 0, len(r.Ports))
	for _, p := range r.Ports {
		parts = append(parts, net.JoinHostPort(host, strconv.Itoa(p)))
	}
	return strings.Join(parts, ",")
}

File: crates/hive-egress/src/allowlist.rs
//! What a run may reach.

use std::fmt;
use std::net::{IpAddr, Ipv4Addr};

/// The ports an entry without one permits. Deliberately short: an agent that
/// needs 8443 should say 8443.
pub const DEFAULT_PORTS: [u16; 2] = [80, 443];

/// Marks an entry as permitted to resolve to a non-public address:
/// `private:printer.home.example.com`. Spelled out rather than inferred,
/// because "this name is allowed to point inside the network" is exactly the
/// sentence someone should have to write down.
pub const PRIVATE_PREFIX: &str = "private:";

/// One allowlist entry.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Rule {
    /// A hostname, an IP literal, or (with `wildcard`) the dotted suffix of a
    /// `*.` form.
    pub host: String,
    /// Ports the rule permits. Empty means `DEFAULT_PORTS`.
    pub ports: Vec<u16>,
    wildcard: bool,
    /// Lets THIS rule reach a non-public address. Per rule, not per
    /// allowlist: a global flag meant that allowing one LAN printer widened
    /// the SSRF guard for every other entry.
    allow_private: bool,
    /// Set when `host` parsed as an IP literal, so an explicit address entry
    /// matches without a DNS round trip.
    ip: Option<IpAddr>,
}

impl Rule {
    fn permits_port(&self, port: u16) -> bool {
        if self.ports.is_empty() {
            DEFAULT_PORTS.contains(&port)
        } else {
            self.ports.contains(&port)
        }
    }

    fn matches_host(&self, host: &str) -> bool {
        if self.wildcard {
            // "*.example.com" covers sub.example.com and deeper, and
            // deliberately not example.com. Allowing the apex by implication
            // is how an allowlist ends up broader than the person writing it
            // believed.
            return host.ends_with(&self.host);
        }
        if let Some(ip) = self.ip {
            return host.parse::<IpAddr>().is_ok_and(|c| same_ip(ip, c));
        }
        host == self.host
    }
}

impl fmt::Display for Rule {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let host = if self.wildcard {
            format!("*{}", self.host)
        } else {
            self.host.clone()
        };
        if self.ports.is_empty() {
            return f.write_str(&host);
        }
        let parts: Vec<String> = self
            .ports
            .iter()
            .map(|p| join_host_port(&host, *p))
            .collect();
        f.write_str(&parts.join(","))
    }
}

fn join_host_port(host: &str, port: u16) -> String {
    if host.contains(':') {
        format!("[{host}]:{port}")
    } else {
        format!("{host}:{port}")
    }
}

/// A 4-byte address and its 4-in-6 form are the same address.
fn same_ip(a: IpAddr, b: IpAddr) -> bool {
    canonical(a) == canonical(b)
}

fn canonical(ip: IpAddr) -> IpAddr {
    match ip {
        IpAddr::V6(v6) => v6
            .to_ipv4_mapped()
            .map(IpAddr::V4)
            .unwrap_or(IpAddr::V6(v6)),
        v4 => v4,
    }
}

#[derive(Debug, Clone, PartialEq, Eq, thiserror::Error)]
#[error("allowlist entry {entry:?}: {reason}")]
pub struct AllowlistError {
    pub entry: String,
    pub reason: String,
}

/// A refusal by policy, as opposed to a network failure. The distinction is
/// load-bearing: a policy denial is a 403 and means the allowlist did its
/// job, while an unreachable upstream is a 502 and means something is broken.
#[derive(Debug, Clone, PartialEq, Eq, thiserror::Error)]
#[error("{0}")]
pub struct Denied(pub String);

/// Decides what a run may reach. The default denies everything, which is the
/// point: a proxy that starts up empty is safe, not broken.
#[derive(Clone, Debug, Default)]
pub struct Allowlist {
    rules: Vec<Rule>,
    /// Permits RFC1918, loopback, link-local and ULA destinations for every
    /// rule. Off by default. This is the DNS-rebinding and SSRF control:
    /// without it, allowlisting `metrics.example.com` is enough to reach
    /// 169.254.169.254, because the attacker controls what the name resolves
    /// to. Turning it on is a deliberate widening; the per-rule prefix is the
    /// narrow alternative.
    pub allow_private_destinations: bool,
}

/// Builds an allowlist from entries.
///
/// ```text
/// example.com            host on ports 80 and 443
/// example.com:8443       host on one port
/// *.example.com          subdomains of example.com, NOT example.com itself
/// 192.0.2.10             an address literal
/// private:printer.lan    a name that may resolve somewhere private
/// ```
///
/// An empty list is valid and denies everything.
pub fn parse_allowlist<S: AsRef<str>>(entries: &[S]) -> Result<Allowlist, AllowlistError> {
    let mut list = Allowlist::default();
    for raw in entries {
        let entry = raw.as_ref().trim();
        if entry.is_empty() {
            continue;
        }
        let rule = parse_rule(entry).map_err(|reason| AllowlistError {
            entry: raw.as_ref().to_string(),
            reason,
        })?;
        list.rules.push(rule);
    }
    Ok(list)
}

/// Splits `host[:port]`, `[v6]:port` and bare `[v6]` forms. `None` port when
/// there is none to split off.
fn split_host_port(entry: &str) -> Result<(String, Option<u16>), String> {
    if let Some(rest) = entry.strip_prefix('[') {
        let Some((host, after)) = rest.split_once(']') else {
            return Err("malformed bracketed address".into());
        };
        if after.is_empty() {
            return Ok((host.to_string(), None));
        }
        let Some(port) = after.strip_prefix(':') else {
            return Err("malformed bracketed address".into());
        };
        return Ok((host.to_string(), Some(parse_port(port)?)));
    }
    // More than one colon and no brackets is a bare IPv6 literal, not a port.
    if entry.matches(':').count() == 1
        && let Some((host, port)) = entry.rsplit_once(':')
    {
        return Ok((host.to_string(), Some(parse_port(port)?)));
    }
    Ok((entry.to_string(), None))
}

fn parse_port(p: &str) -> Result<u16, String> {
    let n: u32 = p
        .parse()
        .map_err(|_| format!("port {p:?} is not a number"))?;
    if !(1..=65535).contains(&n) {
        return Err(format!("port {n} is out of range"));
    }
    Ok(n as u16)
}

fn parse_rule(entry: &str) -> Result<Rule, String> {
    let mut allow_private = false;
    let mut entry = entry;
    if let Some(rest) = entry.strip_prefix(PRIVATE_PREFIX) {
        allow_private = true;
        entry = rest.trim();
        if entry.is_empty() {
            return Err(format!("no host after the {PRIVATE_PREFIX:?} prefix"));
        }
    }
    let (host, port) = split_host_port(entry)?;
    let host = host.trim_end_matches('.').to_ascii_lowercase();
    if host.is_empty() {
        return Err("no host".into());
    }
    let ports = port.map(|p| vec![p]).unwrap_or_default();
    let mut rule = Rule {
        host: String::new(),
        ports,
        wildcard: false,
        allow_private,
        ip: None,
    };
    if host == "*" {
        // An allowlist that allows everything is almost certainly a mistake,
        // and if it is not, it should be spelled out host by host.
        return Err("\"*\" would allow every host; list them explicitly".into());
    }
    if let Some(suffix) = host.strip_prefix("*.") {
        if suffix.is_empty() || suffix.contains('*') {
            return Err("malformed wildcard".into());
        }
        rule.host = format!(".{suffix}");
        rule.wildcard = true;
        return Ok(rule);
    }
    if host.contains('*') {
        return Err("a wildcard is only supported as a leading \"*.\" label".into());
    }
    rule.ip = host.parse::<IpAddr>().ok();
    // Naming a private address literally IS the opt-in. Writing 192.168.1.50
    // in an allowlist and having it silently never match was the other half
    // of this bug: the entry looked effective and was inert.
    if rule.ip.is_some_and(is_private) {
        rule.allow_private = true;
    }
    rule.host = host;
    Ok(rule)
}

impl Allowlist {
    pub fn rules(&self) -> &[Rule] {
        &self.rules
    }

    pub fn is_empty(&self) -> bool {
        self.rules.is_empty()
    }

    /// Whether host:port is allowed by name. It says nothing about where the
    /// name resolves; `permits_addr` is the other half.
    pub fn permits(&self, host: &str, port: u16) -> bool {
        self.match_rule(host, port).is_some()
    }

    /// The rule that permits host:port, if any. The rule matters to the
    /// second check: whether a resolved address may be dialled is a question
    /// about the entry that allowed the name, not about the list as a whole.
    pub fn match_rule(&self, host: &str, port: u16) -> Option<&Rule> {
        let host = host.trim_end_matches('.').to_ascii_lowercase();
        let host = host.trim_start_matches('[').trim_end_matches(']');
        self.rules
            .iter()
            .find(|r| r.permits_port(port) && r.matches_host(host))
    }

    /// Whether a resolved address may be dialled under a rule. Separate from
    /// `permits` because the two failures are different: a name that is not
    /// on the list is a misconfiguration, and a name on the list that resolves
    /// somewhere private is an attack or an accident that looks like one.
    pub fn permits_addr(&self, rule: &Rule, ip: IpAddr) -> Result<(), Denied> {
        if !is_private(ip) {
            return Ok(());
        }
        if self.allow_private_destinations || rule.allow_private {
            return Ok(());
        }
        Err(Denied(format!("destination {ip} is not a public address")))
    }
}

/// Everything a run has no business reaching by resolving a public name:
/// loopback, RFC1918, link-local (where cloud metadata services live), ULA,
/// carrier-grade NAT (also container runtimes), multicast and the
/// unspecified address. A 4-in-6 address is judged as its 4-byte self.
pub fn is_private(ip: IpAddr) -> bool {
    match canonical(ip) {
        IpAddr::V4(v4) => {
            v4.is_loopback()
                || v4.is_private()
                || v4.is_unspecified()
                || v4.is_link_local()
                || v4.is_multicast()
                || v4.is_broadcast()
                || is_cgnat(v4)
        }
        IpAddr::V6(v6) => {
            v6.is_loopback()
                || v6.is_unspecified()
                || v6.is_multicast()
                || v6.is_unicast_link_local()
                || v6.is_unique_local()
        }
    }
}

/// 100.64.0.0/10.
fn is_cgnat(v4: Ipv4Addr) -> bool {
    let o = v4.octets();
    o[0] == 100 && (64..=127).contains(&o[1])
}
